<a href="https://colab.research.google.com/github/eternitvy/ML_26/blob/main/JS05/JS05_TUGAS2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# Import libary
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
# Preprocessing

# Memuat dataset (gunakan encoding latin-1 khusus untuk data spam.csv)
df = pd.read_csv('spam.csv', encoding='latin-1')

# Membuang kolom yang tidak digunakan (Unnamed: 2, Unnamed: 3, Unnamed: 4)
df = df.drop(df.columns[[2, 3, 4]], axis=1)

# Mengganti nama kolom agar lebih mudah dipahami
df.columns = ['Labels', 'SMS']

# Melakukan encoding pada label (ham = 0, spam = 1)
df['Labels'] = df['Labels'].map({'spam': 1, 'ham': 0})

# Memisahkan fitur (X) dan target (y)
X = df['SMS']
y = df['Labels']

# Membagi data menjadi data latih (80%) dan data uji (20%)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Data berhasil dimuat dan dipersiapkan!")
display(df.head())

Data berhasil dimuat dan dipersiapkan!


,Labels,SMS
0,0,"Go until jurong point, crazy.. Available only ..."
1,0,Ok lar... Joking wif u oni...
2,1,Free entry in 2 a wkly comp to win FA Cup fina...
3,0,U dun say so early hor... U c already then say...
4,0,"Nah I don't think he goes to usf, he lives aro..."


In [3]:
# Model Multinomial naive bayes dengan countvectorizer
# 1. Inisialisasi CountVectorizer dengan stop_words
cv = CountVectorizer(stop_words='english')

# 2. Transformasi teks menjadi matriks angka
X_train_cv = cv.fit_transform(X_train)
X_test_cv = cv.transform(X_test)

# 3. Pembuatan dan Pelatihan Model
mnb_cv = MultinomialNB()
mnb_cv.fit(X_train_cv, y_train)

# 4. Prediksi dan Evaluasi
y_pred_cv = mnb_cv.predict(X_test_cv)
acc_cv = accuracy_score(y_test, y_pred_cv)

print("=== EVALUASI MODEL COUNTVECTORIZER ===")
print(f"Akurasi: {acc_cv * 100:.2f}%")
print("\nClassification Report:\n", classification_report(y_test, y_pred_cv))

=== EVALUASI MODEL COUNTVECTORIZER ===
Akurasi: 98.39%

Classification Report:
               precision    recall  f1-score   support

           0       0.99      0.99      0.99       965
           1       0.96      0.92      0.94       150

    accuracy                           0.98      1115
   macro avg       0.97      0.96      0.96      1115
weighted avg       0.98      0.98      0.98      1115



In [4]:
# Model Multinomial naive bayes dengan TF-IDF
# 1. Inisialisasi TfidfVectorizer dengan stop_words
tfidf = TfidfVectorizer(stop_words='english')

# 2. Transformasi teks menjadi matriks TF-IDF
X_train_tfidf = tfidf.fit_transform(X_train)
X_test_tfidf = tfidf.transform(X_test)

# 3. Pembuatan dan Pelatihan Model
mnb_tfidf = MultinomialNB()
mnb_tfidf.fit(X_train_tfidf, y_train)

# 4. Prediksi dan Evaluasi
y_pred_tfidf = mnb_tfidf.predict(X_test_tfidf)
acc_tfidf = accuracy_score(y_test, y_pred_tfidf)

print("=== EVALUASI MODEL TF-IDF ===")
print(f"Akurasi: {acc_tfidf * 100:.2f}%")
print("\nClassification Report:\n", classification_report(y_test, y_pred_tfidf))

=== EVALUASI MODEL TF-IDF ===
Akurasi: 96.68%

Classification Report:
               precision    recall  f1-score   support

           0       0.96      1.00      0.98       965
           1       1.00      0.75      0.86       150

    accuracy                           0.97      1115
   macro avg       0.98      0.88      0.92      1115
weighted avg       0.97      0.97      0.96      1115



**ANALISA**

Akurasi TF-IDF menghasilkan akurasi sedikit lebih rendah, yaitu 96.68%. Nilai recall untuk kelas spam (kelas 1) turun drastis menjadi hanya 75%. Artinya, ada lebih banyak pesan spam yang lolos (terprediksi sebagai pesan biasa/ham) dibandingkan saat menggunakan CountVectorizer.

**KESIMPULAN**

Pada kasus deteksi spam SMS ini, fitur CountVectorizer (Tugas 2) adalah yang TERBAIK.

Karena algoritma Multinomial Naive Bayes dirancang untuk bekerja pada data yang berupa jumlah kemunculan diskrit (bilangan bulat seperti 1, 2, 3 kali muncul) yang dihasilkan oleh CountVectorizer. Sementara TF-IDF menghasilkan nilai pecahan/desimal yang secara teori dasar melanggar asumsi distribusi multinomial, meskipun pada praktiknya masih bisa dijalankan.